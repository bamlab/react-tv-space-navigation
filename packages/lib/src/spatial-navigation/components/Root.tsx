import { ReactNode, useRef } from 'react';
import { ParentIdContext } from '../context/ParentIdContext';
import { SpatialNavigatorContext } from '../context/SpatialNavigatorContext';
import { useBeforeMountEffect } from '../hooks/useBeforeMountEffect';
import { useCreateSpatialNavigator } from '../hooks/useCreateSpatialNavigator';
import { useRemoteControl } from '../hooks/useRemoteControl';
import { OnDirectionHandledWithoutMovement } from '../SpatialNavigator';

const ROOT_ID = 'root';

type Props = {
  /**
   * Determines if the spatial navigation is active.
   * If false, the spatial navigation will be locked, and no nodes can be focused.
   * This is useful to handle a multi page app: you can disable the non-focused pages' spatial navigation roots.
   */
  isActive?: boolean;
  /**
   * Called when you're reaching a border of the navigator.
   * A use case for this would be the implementation of a side menu
   * that's shared between pages. You can have a separate navigator
   * for your side menu, which would be common across pages, and you'd
   * make this menu active when you reach the left side of your page navigator.
   */
  onDirectionHandledWithoutMovement?: OnDirectionHandledWithoutMovement;
  children: ReactNode;
};

export const SpatialNavigationRoot = ({
  isActive = true,
  onDirectionHandledWithoutMovement = () => undefined,
  children,
}: Props) => {
  // We can't follow the react philosophy here: we can't recreate a navigator if this function changes
  // so we'll have to store its ref and update the ref if there is a new value to this function
  const onDirectionHandledWithoutMovementRef = useRef<OnDirectionHandledWithoutMovement>(
    () => undefined,
  );
  // Update the ref at every render
  onDirectionHandledWithoutMovementRef.current = onDirectionHandledWithoutMovement;

  const spatialNavigator = useCreateSpatialNavigator({
    onDirectionHandledWithoutMovementRef,
  });

  useRemoteControl({ spatialNavigator, isActive });

  useBeforeMountEffect(() => {
    spatialNavigator.registerNode(ROOT_ID, { orientation: 'vertical' });
    return () => spatialNavigator.unregisterNode(ROOT_ID);
  }, []);

  return (
    <SpatialNavigatorContext.Provider value={spatialNavigator}>
      <ParentIdContext.Provider value={ROOT_ID}>{children}</ParentIdContext.Provider>
    </SpatialNavigatorContext.Provider>
  );
};
