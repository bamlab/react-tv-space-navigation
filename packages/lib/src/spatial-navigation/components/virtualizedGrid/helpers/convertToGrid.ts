import chunk from 'lodash/chunk';
import { GridRowType } from '../SpatialNavigationVirtualizedGrid';
import { ItemWithIndex } from '../../virtualizedList/VirtualizedList';
import { NodeOrientation } from '../../../types/orientation';

export const convertToGrid = <T extends ItemWithIndex>(
  data: T[],
  numberOfColumns: number,
  header?: JSX.Element,
): GridRowType<T>[] => {
  const rows: T[][] = chunk(data, numberOfColumns);

  return rows.map((items, index) => {
    //We do this to have index taking into account the header
    const computedIndex = header ? index + 1 : index;
    return { items, index: computedIndex };
  });
};

export const invertOrientation = (orientation: NodeOrientation): NodeOrientation =>
  orientation === 'vertical' ? 'horizontal' : 'vertical';
