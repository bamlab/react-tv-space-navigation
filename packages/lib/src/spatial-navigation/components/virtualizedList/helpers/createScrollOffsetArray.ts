import { ScrollBehavior } from '../VirtualizedList';
import { computeTranslation } from './computeTranslation';
import { getLastLeftItemIndex, getLastRightItemIndex } from './getLastItemIndex';

/**
 * This function precomputes all scroll offsets
 * It won't move until data moves or the itemSize changes
 */
export const computeAllScrollOffsets = <T>({
  itemSize,
  nbMaxOfItems,
  numberOfItemsVisibleOnScreen,
  scrollBehavior,
  data,
  listSizeInPx,
}: {
  itemSize: number | ((item: T) => number);
  nbMaxOfItems: number;
  numberOfItemsVisibleOnScreen: number;
  scrollBehavior: ScrollBehavior;
  data: T[];
  listSizeInPx: number;
}) => {
  const maxPossibleLeftAlignedIndex = getLastLeftItemIndex<T>(data, itemSize, listSizeInPx);
  const maxPossibleRightAlignedIndex = getLastRightItemIndex<T>(data, itemSize, listSizeInPx);

  const scrollOffsets = data.map((_, index) =>
    computeTranslation({
      currentlyFocusedItemIndex: index,
      itemSizeInPx: itemSize,
      nbMaxOfItems: nbMaxOfItems ?? data.length,
      numberOfItemsVisibleOnScreen: numberOfItemsVisibleOnScreen,
      scrollBehavior: scrollBehavior,
      data: data,
      listSizeInPx: listSizeInPx,
      maxPossibleLeftAlignedIndex: maxPossibleLeftAlignedIndex,
      maxPossibleRightAlignedIndex: maxPossibleRightAlignedIndex,
    }),
  );

  return scrollOffsets;
};
