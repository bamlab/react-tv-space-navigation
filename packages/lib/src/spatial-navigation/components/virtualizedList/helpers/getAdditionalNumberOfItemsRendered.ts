import { ScrollBehavior } from '../VirtualizedList';

export const getAdditionalNumberOfItemsRendered = (
  scrollBehavior: ScrollBehavior,
  numberOfElementsVisibleOnScreen: number,
) => {
  switch (scrollBehavior) {
    case 'stick-to-start':
    case 'stick-to-end':
      return 4 + numberOfElementsVisibleOnScreen;
    case 'jump-on-scroll':
      return 2 * numberOfElementsVisibleOnScreen + 1;
  }
};
