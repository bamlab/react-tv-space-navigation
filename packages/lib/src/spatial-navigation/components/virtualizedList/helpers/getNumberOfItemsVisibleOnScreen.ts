const getMinSizeOfItems = <T>({
  data,
  itemSize,
}: {
  data: T[];
  itemSize: number | ((item: T) => number);
}) => {
  if (typeof itemSize === 'number') {
    return itemSize;
  }

  if (data.length === 0) {
    return 0;
  }

  const firstElementSize = itemSize(data[0]);

  const minSize = data.reduce<number>((smallestSize, item) => {
    const currentSize = itemSize(item);
    if (currentSize < smallestSize) return currentSize;
    return smallestSize;
  }, firstElementSize);

  if (minSize === 0) {
    console.warn('The size of the smallest item in the list is 0. The list will appear empty.');
  }

  return minSize;
};

export const getNumberOfItemsVisibleOnScreen = <T>({
  data,
  listSizeInPx,
  itemSize,
}: {
  data: T[];
  listSizeInPx: number;
  itemSize: number | ((item: T) => number);
}) => {
  if (data.length === 0) {
    return 0;
  }

  const itemSizeToComputeRanges = getMinSizeOfItems({ data, itemSize });

  if (!itemSizeToComputeRanges) {
    return 0;
  }

  if (itemSizeToComputeRanges === 0) {
    console.warn('The size of the smallest item in the list is 0. The list will appear empty.');
    return 0;
  }

  return Math.floor(listSizeInPx / itemSizeToComputeRanges);
};
