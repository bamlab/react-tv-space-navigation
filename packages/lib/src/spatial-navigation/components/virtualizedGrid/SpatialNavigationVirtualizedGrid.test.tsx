import { RenderResult, act, render, screen } from '@testing-library/react-native';
import { TestButton } from '../tests/TestButton';
import { SpatialNavigationRoot } from '../Root';
import { ReactTestInstance } from 'react-test-renderer';
import '../tests/helpers/configureTestRemoteControl';

import { DefaultFocus } from '../../context/DefaultFocusContext';
import { SpatialNavigationVirtualizedGrid } from '../virtualizedGrid/SpatialNavigationVirtualizedGrid';
import testRemoteControlManager from '../tests/helpers/testRemoteControlManager';
import { setComponentLayoutSize } from '../../../testing/setComponentLayoutSize';

export const expectButtonToHaveFocus = (component: RenderResult, text: string) => {
  const element = component.getByRole('button', { name: text });
  expect(element).toBeSelected();
};

const expectListToHaveScroll = (listElement: ReactTestInstance, scrollValue: number) =>
  expect(listElement).toHaveStyle({ transform: [{ translateY: scrollValue }] });

describe('SpatialNavigationVirtualizedGrid', () => {
  const renderItem = ({ item, index }: { item: { onSelect: () => void }; index: number }) => (
    <TestButton title={`button ${index + 1}`} onSelect={item.onSelect} />
  );

  function createDataArray(numberOfItems: number) {
    const data = [];
    for (let i = 0; i < numberOfItems; i++) {
      data.push({ onSelect: () => undefined, index: i });
    }
    return data;
  }

  const gridTestId = 'test-grid';

  const renderGrid = () =>
    render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedGrid
            renderItem={renderItem}
            data={createDataArray(28)}
            itemHeight={100}
            numberOfColumns={3}
            testID="test-grid"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );

  it('renders the correct number of item', () => {
    const component = renderGrid();
    setComponentLayoutSize(gridTestId, component, { width: 300, height: 300 });

    expect(screen).toMatchSnapshot();

    const listElement = component.getByTestId(gridTestId);
    expect(listElement).toHaveStyle({ height: 1000 });

    expect(screen.getByText('button 1')).toBeTruthy();
    expectButtonToHaveFocus(component, 'button 1');
    expect(screen.getByText('button 2')).toBeTruthy();
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 4')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.getByText('button 8')).toBeTruthy();
    expect(screen.getByText('button 9')).toBeTruthy();
    expect(screen.getByText('button 10')).toBeTruthy();
    expect(screen.getByText('button 11')).toBeTruthy();
    expect(screen.getByText('button 12')).toBeTruthy();
    expect(screen.getByText('button 13')).toBeTruthy();
    expect(screen.getByText('button 14')).toBeTruthy();
    expect(screen.getByText('button 15')).toBeTruthy();
    expect(screen.getByText('button 16')).toBeTruthy();
    expect(screen.getByText('button 17')).toBeTruthy();
    expect(screen.getByText('button 18')).toBeTruthy();
    expect(screen.getByText('button 19')).toBeTruthy();
    expect(screen.getByText('button 20')).toBeTruthy();
    expect(screen.getByText('button 21')).toBeTruthy();
    expect(screen.queryByText('button 22')).toBeFalsy();
  });

  it('handles correctly RIGHT & DOWN and RENDERS new elements accordingly while deleting elements that are too far from scroll when on stick to start scroll', () => {
    const component = renderGrid();
    setComponentLayoutSize(gridTestId, component, { width: 300, height: 300 });
    act(() => jest.runAllTimers());

    const listElement = component.getByTestId(gridTestId);
    expectListToHaveScroll(listElement, 0);
    expect(listElement).toHaveStyle({ height: 1000 });

    testRemoteControlManager.handleRight();

    expect(screen.getByText('button 1')).toBeTruthy();
    expect(screen.getByText('button 2')).toBeTruthy();
    expectButtonToHaveFocus(component, 'button 2');
    expect(screen.getByText('button 3')).toBeTruthy();
    expect(screen.getByText('button 4')).toBeTruthy();
    expect(screen.getByText('button 5')).toBeTruthy();
    expect(screen.getByText('button 6')).toBeTruthy();
    expect(screen.getByText('button 7')).toBeTruthy();
    expect(screen.getByText('button 8')).toBeTruthy();
    expect(screen.getByText('button 9')).toBeTruthy();
    expect(screen.getByText('button 10')).toBeTruthy();
    expect(screen.getByText('button 11')).toBeTruthy();
    expect(screen.getByText('button 12')).toBeTruthy();
    expect(screen.getByText('button 13')).toBeTruthy();
    expect(screen.getByText('button 14')).toBeTruthy();
    expect(screen.getByText('button 15')).toBeTruthy();
    expect(screen.getByText('button 16')).toBeTruthy();
    expect(screen.getByText('button 17')).toBeTruthy();
    expect(screen.getByText('button 18')).toBeTruthy();
    expect(screen.getByText('button 19')).toBeTruthy();
    expect(screen.getByText('button 20')).toBeTruthy();
    expect(screen.getByText('button 21')).toBeTruthy();
    expect(screen.queryByText('button 22')).toBeFalsy();

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -100);
    expectButtonToHaveFocus(component, 'button 5');

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -200);
    expectButtonToHaveFocus(component, 'button 8');

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -300);
    expectButtonToHaveFocus(component, 'button 11');

    expect(screen.queryByText('button 1')).toBeFalsy();
    expect(screen.queryByText('button 2')).toBeFalsy();
    expect(screen.queryByText('button 3')).toBeFalsy();
    expect(screen.getByText('button 4')).toBeTruthy();

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -400);
    expectButtonToHaveFocus(component, 'button 14');

    expect(screen.queryByText('button 4')).toBeFalsy();
    expect(screen.queryByText('button 5')).toBeFalsy();
    expect(screen.queryByText('button 6')).toBeFalsy();
    expect(screen.getByText('button 7')).toBeTruthy();

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -500);
    expectButtonToHaveFocus(component, 'button 17');

    expect(screen.queryByText('button 7')).toBeFalsy();
    expect(screen.queryByText('button 8')).toBeFalsy();
    expect(screen.queryByText('button 9')).toBeFalsy();
    expect(screen.getByText('button 10')).toBeTruthy();

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -600);
    expectButtonToHaveFocus(component, 'button 20');

    expect(screen.queryByText('button 7')).toBeFalsy();
    expect(screen.queryByText('button 8')).toBeFalsy();
    expect(screen.queryByText('button 9')).toBeFalsy();
    expect(screen.getByText('button 10')).toBeTruthy();

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -700);
    expectButtonToHaveFocus(component, 'button 23');

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -700);
    expectButtonToHaveFocus(component, 'button 26');

    testRemoteControlManager.handleDown();
    expectListToHaveScroll(listElement, -700);
    expectButtonToHaveFocus(component, 'button 28');
  });

  it('handles correctly RIGHT & DOWN and RENDERS new elements accordingly while deleting elements that are too far from scroll when on stick to end scroll', () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedGrid
            renderItem={renderItem}
            data={createDataArray(19)}
            itemHeight={100}
            numberOfColumns={3}
            testID="test-grid"
            scrollBehavior="stick-to-end"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    setComponentLayoutSize(gridTestId, component, { width: 300, height: 300 });
    act(() => jest.runAllTimers());

    const listElement = component.getByTestId(gridTestId);
    expectListToHaveScroll(listElement, 0);
    expect(listElement).toHaveStyle({ height: 700 });

    testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 5');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 8');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 11');
    expectListToHaveScroll(listElement, -100);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 14');
    expectListToHaveScroll(listElement, -200);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 17');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 19');
    expectListToHaveScroll(listElement, -400);
  });

  it('handles correctly RIGHT & DOWN and RENDERS new elements accordingly while deleting elements that are too far from scroll when on jump on scroll scroll', () => {
    const component = render(
      <SpatialNavigationRoot>
        <DefaultFocus>
          <SpatialNavigationVirtualizedGrid
            renderItem={renderItem}
            data={createDataArray(19)}
            itemHeight={100}
            numberOfColumns={3}
            testID="test-grid"
            scrollBehavior="jump-on-scroll"
          />
        </DefaultFocus>
      </SpatialNavigationRoot>,
    );
    setComponentLayoutSize(gridTestId, component, { width: 300, height: 300 });
    act(() => jest.runAllTimers());

    const listElement = component.getByTestId(gridTestId);
    expectListToHaveScroll(listElement, 0);
    expect(listElement).toHaveStyle({ height: 700 });

    testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 5');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 8');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 11');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 14');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 17');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleDown();
    expectButtonToHaveFocus(component, 'button 19');
    expectListToHaveScroll(listElement, -400);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 16');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 13');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 10');
    expectListToHaveScroll(listElement, -300);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 7');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 4');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleUp();
    expectButtonToHaveFocus(component, 'button 1');
    expectListToHaveScroll(listElement, 0);

    testRemoteControlManager.handleRight();
    expectButtonToHaveFocus(component, 'button 2');
    expectListToHaveScroll(listElement, 0);
  });
});
