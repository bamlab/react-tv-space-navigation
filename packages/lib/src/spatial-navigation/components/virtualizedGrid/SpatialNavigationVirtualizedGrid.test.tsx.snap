// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpatialNavigationVirtualizedGrid renders the correct number of item 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "height": "100%",
      "width": "100%",
    }
  }
  testID="test-grid-size-giver"
>
  <View
    collapsable={false}
    style={
      {
        "flex": 1,
        "flexDirection": "column",
        "height": 1000,
        "transform": [
          {
            "translateY": 0,
          },
        ],
      }
    }
    testID="test-grid"
  >
    <View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 1"
              accessibilityLabel="button 1"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": true,
                }
              }
              accessible={true}
              isFocused={true}
              style={
                [
                  {
                    "backgroundColor": "red",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 1
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 2"
              accessibilityLabel="button 2"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 2
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 3"
              accessibilityLabel="button 3"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 3
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 100,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 4"
              accessibilityLabel="button 4"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 4
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 5"
              accessibilityLabel="button 5"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 5
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 6"
              accessibilityLabel="button 6"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 6
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 200,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 7"
              accessibilityLabel="button 7"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 7
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 8"
              accessibilityLabel="button 8"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 8
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 9"
              accessibilityLabel="button 9"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 9
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 300,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 10"
              accessibilityLabel="button 10"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 10
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 11"
              accessibilityLabel="button 11"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 11
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 12"
              accessibilityLabel="button 12"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 12
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 400,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 13"
              accessibilityLabel="button 13"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 13
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 14"
              accessibilityLabel="button 14"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 14
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 15"
              accessibilityLabel="button 15"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 15
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 500,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 16"
              accessibilityLabel="button 16"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 16
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 17"
              accessibilityLabel="button 17"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 17
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 18"
              accessibilityLabel="button 18"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 18
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "left": 0,
            "position": "absolute",
            "transform": [
              {
                "translateY": 600,
              },
            ],
          }
        }
      >
        <View
          style={
            [
              undefined,
              {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View>
            <View
              accessibilityHint="button 19"
              accessibilityLabel="button 19"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 19
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 20"
              accessibilityLabel="button 20"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 20
              </Text>
            </View>
          </View>
          <View>
            <View
              accessibilityHint="button 21"
              accessibilityLabel="button 21"
              accessibilityRole="button"
              accessibilityState={
                {
                  "selected": false,
                }
              }
              accessible={true}
              isFocused={false}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text>
                button 21
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
